File: src/java/foundgine-runtime/src/main/java/com/foundgine/runtime/routing/DefaultRoutingEngine.java
package com.foundgine.runtime.routing;

import java.util.*;

public final class DefaultRoutingEngine implements IRoutingEngine {
	private final List<IRoutingRule> rules;

	public DefaultRoutingEngine(Collection<? extends IRoutingRule> rules) {
		this.rules = List.copyOf(rules == null ? List.of() : rules);
	}

	public DefaultRoutingEngine() {
		this(null);
	}

	public TaskContract route(RoutingContext context) {
		Objects.requireNonNull(context, "context");
		for (var rule : rules) {
			var d = rule.tryRoute(context);
			if (d != null && d.matched() && d.contract() != null)
				return d.contract();
		}
		return TaskContract.defaults(UUID.randomUUID().toString().replace("-", ""));
	}
}

File: src/csharp/Foundgine.Runtime/Routing/DefaultRoutingEngine.cs
namespace Foundgine.Runtime.Routing;

/// <summary>
/// Evaluates <see cref="IRoutingRule"/>s in order and returns the first
/// non-abstaining result. Falls back to <see cref="TaskContract.Default"/>
/// (foreground, local, new) when no rule matches, so an empty or
/// misconfigured rule set never blocks execution — it just declines to
/// specialize it.
/// </summary>
public sealed class DefaultRoutingEngine : IRoutingEngine
{
    private readonly IReadOnlyList<IRoutingRule> _rules;

    public DefaultRoutingEngine(IEnumerable<IRoutingRule>? rules = null)
    {
        _rules = rules?.ToArray() ?? Array.Empty<IRoutingRule>();
    }

    public TaskContract Route(RoutingContext context)
    {
        ArgumentNullException.ThrowIfNull(context);

        foreach (var rule in _rules)
        {
            if (rule.TryRoute(context, out var contract) && contract is not null)
                return contract;
        }

        return TaskContract.Default(taskId: Guid.NewGuid().ToString("n"));
    }
}